src/app/clusters/ota-requestor/OTARequestor.cpp

```diff
@@ -84,84 +55,138 @@ OTARequestorInterface * GetRequestorInstance()
+        return CHIP_ERROR_MISSING_URI_SEPARATOR;
     }
 
-    VerifyOrReturn(mBdxDownloader != nullptr, ChipLogError(SoftwareUpdate, "downloader is null"));
+    // Extract file designator from the path field
+    size_t fileDesignatorLength = uri.size() - (strlen(bdxPrefix) + kNodeIdHexStringMaxLen + strlen(bdxSeparator));
+    fileDesignator              = uri.SubSpan(uri.size() - fileDesignatorLength, fileDesignatorLength);
 
-    // TODO: allow caller to provide their own OTADownloader instance and set BDX parameters
+    return CHIP_NO_ERROR;
+}
+
+void OTARequestor::OnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response)
+{
+    LogQueryImageResponse(response);
 
-    TransferSession::TransferInitData initOptions;
-    initOptions.TransferCtlFlags = chip::bdx::TransferControlFlags::kReceiverDrive;
-    initOptions.MaxBlockSize     = 1024;
-    char testFileDes[9]          = { "test.txt" };
-    initOptions.FileDesLength    = static_cast<uint16_t>(strlen(testFileDes));
-    initOptions.FileDesignator   = reinterpret_cast<uint8_t *>(testFileDes);
+    VerifyOrReturn(context != nullptr, ChipLogError(SoftwareUpdate, "Received QueryImageResponse with invalid context"));
 
-    chip::OperationalDeviceProxy * operationalDeviceProxy = Server::GetInstance().GetOperationalDeviceProxy();
-    if (operationalDeviceProxy != nullptr)
+    OTARequestor * requestorCore = static_cast<OTARequestor *>(context);
+
+    switch (response.status)
     {
-        chip::Messaging::ExchangeManager * exchangeMgr = operationalDeviceProxy->GetExchangeManager();
-        chip::Optional<chip::SessionHandle> session    = operationalDeviceProxy->GetSecureSession();
-        if (exchangeMgr != nullptr && session.HasValue())
-        {
-            exchangeCtx = exchangeMgr->NewContext(session.Value(), &mBdxMessenger);
-        }
+    case EMBER_ZCL_OTA_QUERY_STATUS_UPDATE_AVAILABLE: {
+        VerifyOrReturn(response.imageURI.HasValue(), ChipLogError(SoftwareUpdate, "Update is available but no image URI present"));
 
-        if (exchangeCtx == nullptr)
-        {
-            ChipLogError(BDX, "unable to allocate ec: exchangeMgr=%p sessionExists? %u", exchangeMgr, session.HasValue());
-            return;
-        }
+        // Parse out the provider node ID and file designator from the image URI
+        NodeId nodeId = chip::kUndefinedNodeId;
+        CharSpan fileDesignator;
+        CHIP_ERROR err = requestorCore->ParseBdxUri(response.imageURI.Value(), nodeId, fileDesignator);
+        VerifyOrReturn(err == CHIP_NO_ERROR,
+                       ChipLogError(SoftwareUpdate, "BDX image URI is invalid: %" CHIP_ERROR_FORMAT, err.Format()));
+        requestorCore->mProviderNodeId = nodeId;
+
+        // CSM should already be created for sending QueryImage command so use the same CSM based on the assumption that the
+        // provider node ID that will supply the OTA image is on the same fabric as the sender of the QueryImageResponse
```

Comment:
Can this assumption be validated?

Reply:
This assumption was something I've talked to you about and you said yes...

Reply:
I said that it would likely be the same in most situations. I didn't say we should assume it is outright no matter what :)

Reply:
Sounds like we can start here, and support across fabric later. (I'm honestly not sure how that even works...)

Reply:
@tcarmelveilleux If we can't assume this outright, then the QueryImageResponse should contain the fabric. Otherwise, there's no other way to know.

Reply:
Misunderstanding here:

- The *fabric will be the same*
- The *node ID* may not be

I think the comment can be rewritten as:
```suggestion
        // CSM should already be created for sending QueryImage command so use the same CSM since the
        // provider node ID that will supply the OTA image must be on the same fabric as the sender of the QueryImageResponse
```